File: 백트래킹/알파벳/baekjoon-1987.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int dy[] = {-1, 1, 0, 0};
int dx[] = {0, 0, -1, 1};

int go(vector<string> &board, vector<bool> &check, int y, int x)
{
  int ans = 0;
  for (int i = 0; i < 4; i++)
  {
    int ny = y + dy[i];
    int nx = x + dx[i];

    if (0 <= ny && ny < board.size() && 0 <= nx && nx < board[0].size())
    {
      if (check[board[ny][nx] - 'A'] == false)
      {
        check[board[ny][nx] - 'A'] = true;
        int next = go(board, check, ny, nx);
        if(ans < next) {
          ans = next;
        }
        check[board[ny][nx] - 'A'] = false;
      }
    }
  }
  return ans + 1;
}

int main()
{
  int r, c;
  cin >> r >> c;
  vector<string> board(r);
  for (int i = 0; i < r; i++)
  {
    cin >> board[i];
  }
  vector<bool> check(26);
  check[board[0][0] - 'A'] = true;
  cout << go(board, check, 0, 0) << "\n";
  return 0;
}

File: 백트래킹/알파벳/baekjoon-1987.py
dy, dx = [-1, 1, 0, 0], [0, 0, -1, 1]


def go(y, x):
    ans = 0
    for i in range(4):
        ny, nx = y + dy[i], x + dx[i]
        if 0 <= ny < len(board) and 0 <= nx < len(board[0]):
            index = ord(board[ny][nx]) - 65
            if not check[index]:
                check[index] = True
                next = go(ny, nx)
                ans = max(ans, next)
                check[index] = False
    return ans + 1


r, c = map(int, input().split())
board = [input() for _ in range(r)]
check = [False] * 26
check[ord(board[0][0]) - 65] = True
print(go(0, 0))
